### PR TITLE
Allow passing custom props to the render method

### PR DESCRIPTION
**Pull request by jordan-mcrae**
Allow components to receive optional custom props. This is extremely helpful when components need to access shared data such as data from a hook, or anything that might be needed for additional render logic. For example, theming metadata, additional rendering logic based on a user's authentication status, handling window resize events, etc.

Example usage:

```jsx
const myPortableTextComponents = {
  types: {
    myCustomComponent: ({ children, customProps }) => <p>
      {children}
      {!customProps?.authenticated && <span>A custom message for users who are not logged in...</span>}
    </p>,
  },
}

const YourComponent = (props) => {
  const [authenticated] = useAuthentication();

  return <PortableText
    value={props.value}
    components={myPortableTextComponents}
    customProps={{ authenticated }}
    />
}
```

### CODE REVIEW

**Comment by RobEasthope** [comment]
+1 for this and thanks for putting the PR together @jordan-mcrae

**Comment by stipsan** [comment]
Closing this as the preferred approach is outlined in @snorrees's [reply](https://github.com/portabletext/react-portabletext/pull/17#issuecomment-1479388790).

Another alternative is to use React.createContext yourself, and wrap the provider around `<PortableText>` and have `useContext` in your `types` components.

TL;DR we'd like to keep the API surface, as well as the resulting bundle size, as small as possible. To change the core like this it would have to be justified against those priorities.
